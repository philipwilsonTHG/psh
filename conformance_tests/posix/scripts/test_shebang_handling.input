# Test shebang handling and script execution
# PSH supports script execution with proper interpreter detection

echo "Testing shebang handling:"

# Create test scripts with different shebangs
mkdir -p tmp/shebang_test

# Test basic shell script
cat > tmp/shebang_test/basic.sh << 'EOF'
#!/bin/sh
echo "Basic shell script executed"
echo "Args: $*"
EOF
chmod +x tmp/shebang_test/basic.sh

# Test psh script
cat > tmp/shebang_test/psh_script.sh << 'EOF'
#!/usr/bin/env psh
echo "PSH script executed"
echo "Script name: $0"
echo "First arg: $1"
EOF
chmod +x tmp/shebang_test/psh_script.sh

# Test script without shebang (should use current shell)
cat > tmp/shebang_test/no_shebang.sh << 'EOF'
echo "No shebang script executed"
echo "Current shell handles this"
EOF
chmod +x tmp/shebang_test/no_shebang.sh

# Test python script (to verify interpreter switching)
cat > tmp/shebang_test/python_test.py << 'EOF'
#!/usr/bin/env python3
print("Python script executed")
print("This is from Python")
EOF
chmod +x tmp/shebang_test/python_test.py

# Execute basic shell script
echo "Testing basic shell script:"
tmp/shebang_test/basic.sh arg1 arg2

# Execute script via psh explicitly
echo "Testing psh script execution:"
python3 -m psh tmp/shebang_test/psh_script.sh test_arg

# Execute no-shebang script
echo "Testing no-shebang script:"
tmp/shebang_test/no_shebang.sh

# Test python script execution
echo "Testing Python script:"
if command -v python3 >/dev/null; then
    tmp/shebang_test/python_test.py
else
    echo "Python not available, skipping Python test"
fi

# Test shebang with absolute path
cat > tmp/shebang_test/absolute.sh << 'EOF'
#!/bin/bash
echo "Absolute path shebang"
echo "Bash version info: ${BASH_VERSION:-not bash}"
EOF
chmod +x tmp/shebang_test/absolute.sh

echo "Testing absolute path shebang:"
if command -v bash >/dev/null; then
    tmp/shebang_test/absolute.sh
else
    echo "Bash not available, skipping absolute path test"
fi

# Test script execution error handling
echo "Testing error handling:"
cat > tmp/shebang_test/error_script.sh << 'EOF'
#!/bin/sh
echo "Before error"
false  # This will return exit code 1
echo "After error"
EOF
chmod +x tmp/shebang_test/error_script.sh

tmp/shebang_test/error_script.sh
echo "Error script exit code: $?"

# Test script with complex shebang
cat > tmp/shebang_test/complex.sh << 'EOF'
#!/usr/bin/env sh
# Complex script with comments
echo "Complex script executed"
for i in 1 2 3; do
    echo "Loop iteration: $i"
done
EOF
chmod +x tmp/shebang_test/complex.sh

echo "Testing complex script:"
tmp/shebang_test/complex.sh

# Test script permissions
echo "Testing script permissions:"
cat > tmp/shebang_test/no_exec.sh << 'EOF'
#!/bin/sh
echo "This should not execute without +x"
EOF

# Try to execute without execute permission (should fail)
if tmp/shebang_test/no_exec.sh 2>/dev/null; then
    echo "Script executed without permission (unexpected)"
else
    echo "Script correctly failed without execute permission"
fi

# Cleanup
rm -rf tmp/shebang_test

echo "Shebang handling test completed"