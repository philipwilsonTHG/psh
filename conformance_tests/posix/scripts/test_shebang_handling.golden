Testing shebang handling:
Basic shell script executed
Args:
PSH script executed
Script name: psh
First arg:
No shebang script executed
Current shell handles this
Testing basic shell script:
Testing psh script execution:
Testing no-shebang script:
Testing Python script:
Absolute path shebang
Bash version info: not bash
Testing absolute path shebang:
Testing error handling:
Before error
After error
Error script exit code: 127
Complex script executed
Loop iteration: 1
Loop iteration: 2
Loop iteration: 3
Testing complex script:
Testing script permissions:
This should not execute without +x
Script correctly failed without execute permission
Shebang handling test completed
